pkg/kube/namespace/filter.go

```diff
@@ -266,3 +267,81 @@ func (d *discoveryNamespacesFilter) isSelected(labels labels.Set) bool {
 
 	return false
 }
+
+type maistraDiscoveryNamespacesFilter struct {
+	lock       sync.RWMutex
+	namespaces sets.String
+	handlers   []func(ns string, event model.Event)
+}
+
+func NewMaistraDiscoveryNamespacesFilter(mrc memberroll.MemberRollController) DiscoveryNamespacesFilter {
+	d := &maistraDiscoveryNamespacesFilter{}
+	mrc.Register(d, "MaistraDiscoveryNamespacesFilter")
+	return d
+}
+
+func (d *maistraDiscoveryNamespacesFilter) SetNamespaces(namespaces []string) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.namespaces = sets.New[string](namespaces...)
+}
+
+func (d *maistraDiscoveryNamespacesFilter) Filter(obj interface{}) bool {
+	// xns-informers make sure we only receive objects in member namespaces
+	return true
+}
+
+// FilterNamespace is not implemented, because Maistra does not filter namespaces;
+// this function should never be called if memberRollName is specified, so it's safe to call panic.
+func (d *maistraDiscoveryNamespacesFilter) FilterNamespace(nsMeta metav1.ObjectMeta) bool {
+	panic("Not implemented")
+}
+
+// SelectorsChanged is not implemented, because Maistra does not filter namespaces by selectors;
+// this function should never be called if memberRollName is specified, so it's safe to call panic.
+func (d *maistraDiscoveryNamespacesFilter) SelectorsChanged(discoverySelectors []*metav1.LabelSelector) {
+	log.Warn("Discovery namespaces filter not supported in multi-tenancy. Ignore selector changes.")
```

Comment:
This [line](https://github.com/istio/istio/blob/release-1.20/pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go#L76) will be triggered during the initialization.

Reply:
Ok, thanks

Reply:
Instead of logging this warning, I suggest to change [initMeshWatcherHandler](https://github.com/istio/istio/blob/release-1.20/pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go#L74) as follows and call `panic` here, as it was in previous releases:
```
func (c *Controller) initMeshWatcherHandler(meshWatcher mesh.Watcher, discoveryNamespacesFilter filter.DiscoveryNamespacesFilter) {
	meshWatcher.AddMeshHandler(func() {
		if c.client.GetMemberRollController() != nil {
			discoveryNamespacesFilter.SelectorsChanged(meshWatcher.Mesh().GetDiscoverySelectors())
		}
	})
}
```

Reply:
Done

Reply:
I had to revert the suggested changes as it fails the unit test locally. I will investigate.

Reply:
Found the bug. We should call `SelectedChange` when `c.client.GetMemberRollController() == nil`, not `!=`.